### PR TITLE
[TIR] VNNI and ARM dot product intrinsic for tensorization

### PR DESCRIPTION
**Pull request by masahi**
Introduces a new directory `python/tvm/tir/tensor_intrin` where we put intrinsic descriptions written in TVMScript for various HW targets. They can be used by manual tensorized TIR schedules or auto-tensorized ones. More intrinsics, such as tensorcore ones, DP4A etc will be added later. 

@junrushao1994 @vinx13 @shingjan  @Hzfengsy @spectrometerHBH

### CODE REVIEW

File: python/tvm/tir/tensor_intrin/arm_cpu.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            with T.init():
+                C[i] = T.int32(0)
+            for k in T.serial(0, 4):
+                with T.block("update"):
+                    vi, vk = T.axis.remap("SR", [i, k])
+                    C[vi] = C[vi] + T.cast(A[vk], "int32") * T.cast(B[vi, vk], "int32")
+
+
+@T.prim_func
+def dot_product_4x4_i8i8i32_neon(
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/arm_cpu.py:40]: This is equivalent to the TE one in https://github.com/apache/tvm/blob/912993ff3984ddb47ec98888d645a8f157daa3fc/python/tvm/topi/arm_cpu/tensor_intrin.py#L536 cc @tkonolige

File: python/tvm/tir/tensor_intrin/arm_cpu.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            T.llvm_lookup_intrinsic_id("llvm.aarch64.neon.addp.v4i32"),
+            T.uint32(2),
+            pairwise_reduction_low,
+            pairwise_reduction_high,
+            dtype="int32x4",
+        )
+
+
+@T.prim_func
+def dot_product_4x4_i8i8i32_sdot(
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/arm_cpu.py:101]: This is equivalent to the TE one in https://github.com/apache/tvm/blob/912993ff3984ddb47ec98888d645a8f157daa3fc/python/tvm/topi/arm_cpu/tensor_intrin.py#L431

File: python/tvm/tir/tensor_intrin/x86.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            with T.init():
+                C[i] = T.int32(0)
+            for k in T.serial(0, 4):
+                with T.block("update"):
+                    vi, vk = T.axis.remap("SR", [i, k])
+                    C[vi] = C[vi] + T.cast(A[vk], "int32") * T.cast(B[vi, vk], "int32")
+
+
+@T.prim_func
+def dot_product_16x4_u8i8i32_vnni(
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/x86.py:44]: Equivalent to the TE one in https://github.com/apache/tvm/blob/912993ff3984ddb47ec98888d645a8f157daa3fc/python/tvm/topi/x86/tensor_intrin.py#L244

File: python/tvm/tir/tensor_intrin/arm_cpu.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name
+"""Intrinsics for x86 tensorization."""
+from .. import TensorIntrin
+from tvm.script import tir as T
+
+
+# TODO(masahi): Parametrize the TVMScript description of dot product by
+# shape and dtype, and share the common description with x86.
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/arm_cpu.py:24]: cc @junrushao1994 @yelite, this is one of the common needs for meta programming support in TVMScript. I think shape parameterization is possible via `specialize`, but not sure if I can use that with `T.Buffer` syntax sugar.

File: python/tvm/tir/tensor_intrin/arm_cpu.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        T.writes(C[0:4])
+
+        A_int8 = A.vload([0], "int8x4")
+        re_int32 = T.reinterpret(A_int8, dtype="int32")
+        vec_ai32 = T.broadcast(re_int32, 2)
+        vec_a = T.reinterpret(vec_ai32, dtype="int8x8")
+
+        vec_b = B.vload([0, 0], dtype="int8x16")
+
+        # TODO(masahi): Remove duplication when inlined function call is supported
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/arm_cpu.py:68]: cc @junrushao1994 @yelite, I want to define and call a convenience function like https://github.com/apache/tvm/blob/912993ff3984ddb47ec98888d645a8f157daa3fc/python/tvm/topi/arm_cpu/tensor_intrin.py#L625 to remove duplication in IR generation. I think this is a long-hanging fruit that can be supported without more ground-up work on meta-programming? (just need to tweak the python parser?)

**Comment by junrushao1994** [comment]
CC @vinx13 would you like to review? Thanks a lot!

File: python/tvm/tir/tensor_intrin/arm_cpu.py (hunk 5 above)
Comment (junrushao1994) [reply python/tvm/tir/tensor_intrin/arm_cpu.py:62]: Yes, it is likely doable and @Hzfengsy probably already has something ready

**Review by vinx13 — APPROVED** [review]

**Pull request merged**